### ISSUE TITLE
RFC: JSON Style Guide

### ISSUE THREAD

**Issue by dehora** [zalando/restful-api-guidelines#32]
This issue contains a draft style guide for use with Zalando APIs akin to that provided by [Google](https://gist.github.bus.zalan.do/bdehora/1749ed3c05316c26d2d1), for discussion. It could be a standlone document or a top level section of the existing guidelines (the latter would save having to work with two documents).  There is something to be said for being able to think about media type and JSON content design independent of protocol concerns instead of having them all at the same level. 

If we go ahead with this, the prs would probably be 

 - Create a top level section in the current guidelines and re-organize what's already in place to refactor the document. Existing guidelines will get moved around but won't have their specifications changed.

 - Add in new guidelines as another pr. 

 - See if we want to break the JSON guidelines out to another document or leave them in place.

Note: sections below marked with an asterisk are extracted from the existing guidelines - the exact content below might be out of date with the exact guideline text in master, but that can be fixed up if this becomes a pr. In the meantime it's representative of what things would end up looking like. 

----

# JSON Guidelines

**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Introduction](#introduction)
- [Property Naming](#property-naming)
  - [Must: Property names must be snake_case (and never camelCase)*.](#must-property-names-must-be-snake_case-and-never-camelcase)
  - [Must: Property names must be an ASCII subset](#must-property-names-must-be-an-ascii-subset)
  - [Should: Reserved JavaScript keywords should be avoided](#should-reserved-javascript-keywords-should-be-avoided)
  - [Should: Array names should be pluralized](#should-array-names-should-be-pluralized)
- [Property Values](#property-values)
  - [Must: Boolean property values must not be null](#must-boolean-property-values-must-not-be-null)
  - [Should: Null property values should have their fields removed*](#should-null-property-values-should-have-their-fields-removed)
  - [Should: Empty array values should not be null](#should-empty-array-values-should-not-be-null)
  - [Should: Enumerations should be represented as Strings](#should-enumerations-should-be-represented-as-strings)
  - [Should: Date property values should conform to RFC 3399*](#should-date-property-values-should-conform-to-rfc-3399)
  - [May: Time durations and intervals may conform to ISO 8601](#may-time-durations-and-intervals-may-conform-to-iso-8601)
  - [May: Standards may be used for Country, Language and Currency Codes*](#may-standards-may-be-used-for-country-language-and-currency-codes)
- [Common Objects](#common-objects)
  - [Should: Money should be represented using a common object*](#should-money-should-be-represented-using-a-common-object)
  - [Should: Addresses should be represented using a common object*](#should-addresses-should-be-represented-using-a-common-object)
- [Hypermedia](#hypermedia)
  - [Should: HATEOAS should be used](#should-hateoas-should-be-used)
  - [Should: Standards should be used for Hypermedia Link Types](#should-standards-should-be-used-for-hypermedia-link-types)
  - [Should: Standards should be used for linked/embedded resources](#should-standards-should-be-used-for-linkedembedded-resources)
  - [May: URIs may be used for Custom Link Relations](#may-uris-may-be-used-for-custom-link-relations)

## Introduction
This guide provides recommendations for defining JSON data at Zalando. JSON here refers to [RFC 7159](http://www.rfc-editor.org/rfc/rfc7159.txt), which updates [RFC 4627](https://www.ietf.org/rfc/rfc4627.txt) and the “application/json” media type. This guide clarifies some specific cases to allow Zalando JSON data to have an idiomatic form across teams and services and must be applied to JSON based APIs. 

The keywords "MUST", "MUST NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  and "MAY"  in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

### Must: Property names must be snake_case (and never camelCase)*.
No established industry standard exists, but many popular Internet companies prefer snake_case: e.g. GitHub, Stack Exchange, Twitter. Others, like Google and Amazon, use both - but not only camelCase. It’s essential to establish a consistent look and feel such that JSON looks as if it came from the same hand. 

### Must: Property names must be an ASCII subset
Property names are restricted to ascii encoded strings. The first character must be a letter, an underscore  or a dollar sign, and subsequent characters can be a letter, an underscore, a dollar sign, or a number.

### Should: Reserved JavaScript keywords should be avoided
Most API content is consumed by non-JavaScript clients today, but for security and sanity reasons, JavaScript (strictly ECMAScript) keywords are worth avoiding.

### Should: Array names should be pluralized

To indicate they contain multiple values prefer to pluralize array names. This implies that object names in turn should be singular.

### Must: Boolean property values must not be null
Schema based JSON properties that are by design booleans must not be presented as nulls. A boolean is essentially a closed enumeration of two values, true and false. If the content has a meaningful null value, strongly prefer to replace the boolean with enumeration of named values or statuses - for example accepted_terms_and_conditions with true or false can be replaced with terms_and_conditions with values yes, no and unknown.

### Should: Null property values should have their fields removed*
Swagger/OpenAPI, which is common use, doesn't support null field values (it does allow omitting that field completely if it is not marked as required). However that doesn't prevent clients and servers of sending (and accepting) those fields with null values. 

### Should: Empty array values should not be null
Empty array values can unambiguously be represented as the the empty list, [].

### Should: Enumerations should be represented as Strings
Strings are a reasonable target for values that are by design enumerations. 

### Should: Date property values should conform to RFC 3399*
use the date and time formats defined by [RFC 3339](http://tools.ietf.org/html/rfc3339#section-5.6) — e.g. `2015-05-28T14:09:17+02:00` for a point in time (note that the  [OpenAPI format](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#data-types) "date-time" corresponds to "date-time" in the RFC) and `2015-05-28` for a date (note that the OpenAPI format "date" corresponds to "full-date" in the RFC). Both are specific profiles, a subset of the international standard [ISO 8601](http://en.wikipedia.org/wiki/ISO_8601).

A zone offset may be used -- this is simply defined by the standards. However, we encourage to restrict to UTC (and go without offsets) when dates are passed and returned. From experience we know that zone offsets are not easy to understand and often not correctly handled. Note also, that zone offsets are different from local times that may include daylight saving time.

When it comes to storage, all dates should be consistently stored in UTC without zone offset. Localization should be done locally by the services that provide user interfaces, if required.

Sometimes it can seem data is naturally represented using numerical timestamps, but this can introduce interpretation issues with precision - for example whether to represent a timestamp as 1460062925, 1460062925000 or 1460062925.000. Date strings though more verbose and requiring more effort to parse, lack this ambiguity.

### May: Time durations and intervals may conform to ISO 8601
Schema based JSON properties that are by design durations and intervals may be strings formatted as recommended by ISO 8601.

_note: unsure if we need to spec the form of durations/intervals yet, or we should wait for implementation feedback/need. The duration guideline is taken from google's document._

### May: Standards may be used for Country, Language and Currency Codes*

* [ISO 3166-1-alpha2 country codes](http://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)

* * (It's “GB”, not “UK”, even though “UK” has seen some use at Zalando)

* [ISO 639-1 language code](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)

* * [BCP-47](https://tools.ietf.org/html/bcp47) (based on ISO 639-1) for language variants

* [ISO 4217 currency codes](http://en.wikipedia.org/wiki/ISO_4217)

### Should: Money should be represented using a common object*
Use the following common money structure:

```yaml
Money:
  type: object
  properties:
    amount:
      type: number
      format: decimal
      example: 99.95
    currency:
      type: string
      format: iso-4217
      example: EUR
  required:
    - amount
    - currency
```

Make sure that you don’t convert the “amount” field to `float` / `double` types when implementing
this interface in a specific language or when doing calculations. Otherwise, you might lose
precision. Instead, use exact formats like
Java’s [`BigDecimal`](https://docs.oracle.com/javase/8/docs/api/java/math/BigDecimal.html).
See [Stack Overflow](http://stackoverflow.com/a/3730040/342852) for more info.

Some JSON parsers (NodeJS’s, for example) convert numbers to floats by default. After discussing the
[pros and cons](https://docs.google.com/spreadsheets/d/12wTj-2w39f69XZGwRDrosNc1yWPwQpGgEs_DCt5ODaQ),
we’ve decided on "decimal" as our amount format. It is not a standard OpenAPI format, but should
help us to avoid parsing numbers as float / doubles.

### Should: Addresses should be represented using a common object*
Address structures play a role in different functional and use-case contexts, including country variances. The address structure below should be sufficient for most of our business-related use cases. Use it in your APIs — and compatible extend it if necessary for your API concerns:

```yaml
address:
  description:
    a common address structure adequate for many use cases
  type: object
  properties:
    salutation:
      type: string
      description: |
        A salutation and/or title which may be used for personal contacts. Hint: not to be confused with the gender information that is stored per customer account
      example: Mr
    first_name:
      type: string
      description: given name(s) or first name(s) of a person; may also include middle names
      example: Hans Dieter
    last_name:
      type: string
      description: family name(s) or surname(s) of a person
      example: Mustermann
    business_name:
      type: string
      description: company name of the business organization
      example: Consulting Services GmbH
    street:
      type: string
      description: full street address including house number and street name
      example: Schönhauser Allee 103
    additional:
      type: string
      description: further details like suite, apt no, etc.
      example: 2. Hinterhof rechts
    city:
      type: string
      description: name of the city
      example: Berlin
    zip:
      type: string
      description: Zip code or postal code
      example: 14265
    country_code:
      type: string
      format: iso-3166-1-alpha-2
      example: DE
  required:
    * first_name
    * last_name
    * street
    * city
    * zip
    * country_code
```

### Should: HATEOAS should be used

This is a challenging topic. We don’t have specific recommendations yet; we expect them to emerge
from trial and error, and from the API Guild’s ongoing research. For now, we strongly encourage
teams to apply this principle and document results.

Notes on aspects to be analyzed:

* Security concerns with respect to  passing OAuth2 tokens
* API discovery by client programmers vs. automated hypermedia driven workflows
* Additional client development work vs. reduced efforts in case of API and state handling changes

Additional resources: [Wikipedia](http://en.wikipedia.org/wiki/HATEOAS), [The RESTful CookBook](http://restcookbook.com/Basics/hateoas/)

### Should: Standards should be used for Hypermedia Link Types

To represent hypermedia links in payload results, we should consider using a standard format like [HAL](http://stateless.co/hal_specification.html), [JSON API](http://jsonapi.org/), [JSON-LD](http://json-ld.org/) with [Hydra](http://www.hydra-cg.com/spec/latest/core/), or [Siren](https://github.com/kevinswiber/siren).  (Hint: Read [Kevin Sookocheff’s post](http://sookocheff.com/post/api/on-choosing-a-hypermedia-format/), for instance, to learn more about the hypermedia types.)

### Should: Standards should be used for linked/embedded resources

For HTTP Link headers, consider using a format like [HAL](http://stateless.co/hal_specification.html),[JSON-LD](http://json-ld.org/) with [Hydra](http://www.markus-lanthaler.com/hydra/spec/latest/core/), or [Siren](https://github.com/kevinswiber/siren). For a comparison of these hypermedia types, see
[Kevin Sookocheff’s post](http://sookocheff.com/post/api/on-choosing-a-hypermedia-format/).

### May: URIs may be used for Custom Link Relations

Related or even embedded (sub-) resources should be linked via “meta link attributes” within the response payload; use here the following [HAL](http://stateless.co/hal_specification.html) compliant syntax:

```yaml
    {
      “_links”: {
        “https://docs.team.zalan.do/rels/my-entity”: [{
          “href”: “https://service.team.zalan.do/my-entities/123”
        }]
      }
    }
```

**Comment by mostlymagic** [comment]
JSON guide looks good to me, we should discuss the other changes in a PR

**Comment by fmueller** [comment]
Great idea. I'd also leave it in our api guidelines, just as a separate section. What I still don't understand is, why you have hypermedia section in there. This I'd exclude from json section.

Can you please prepare a pr for it @dehora?

**Comment by dehora** [comment]
> Can you please prepare a pr for it @dehora?

@fmueller 👍 

> What I still don't understand is, why you have hypermedia section in there. 
> This I'd exclude from json section.

Mostly because hypermedia impacts media type designs (eg whether we should use something like HAL). We can go through it in the pr

**Comment by fmueller** [comment]
Actually I'm preparing sth for hypermedia, see #24

**Comment by garrettmac** [comment]
@dehora 
> Must: Property names must be snake_case (and never camelCase)*.
No established industry standard exists, but many popular Internet companies prefer snake_case: e.g. GitHub, Stack Exchange, Twitter. Others, like Google and Amazon, use both - but not only camelCase. It’s essential to establish a consistent look and feel such that JSON looks as if it came from the same hand.

would you mind taking a look at [Google JSON Style Guide](https://google.github.io/styleguide/jsoncstyleguide.xml?showone=Property_Name_Format#Property_Name_Format), where: 


> Property names must be camel-cased, ascii strings.

and updating your issue accordingly? Thanks. 


(Also worth sharing this https://stackoverflow.com/questions/5543490/json-naming-convention)

**Comment by tkrop** [comment]
@garrettmac sorry, this issue and the discussion is closed since 1.5 years. Our rules are based on a very pragmatic decision: We want uniformity and we do not want developers to argue about such topics. That's why we have this rule, and the rest is rolling dices. Sorry, but we will not edit the rule. Anyhow
